### PR TITLE
Use Yup Validations in Bulk New Users

### PR DESCRIPTION
**Pull request by jrjohnson**
Trying this out for the first time on the frontend, it seems to make a few things simpler, which is nice.

### CODE REVIEW

File: packages/frontend/app/components/bulk-new-users.js [hunk 1]
```diff
@@ -182,26 +181,29 @@ export default class BulkNewUsersComponent extends Component {
...
     this.fileUploadError = false;
+    const existingUsernames = await this.getExistingUserNames();
     return new Promise((resolve) => {
       const allowedFileTypes = ['text/plain', 'text/csv', 'text/tab-separated-values'];
       if (!allowedFileTypes.includes(file.type)) {
-        const intl = this.intl;
-        this.set('fileUploadError', true);
-        throw new Error(intl.t('general.fileTypeError', { fileType: file.type }));
+        this.fileUploadError = true;
+        throw new Error(this.intl.t('general.fileTypeError', { fileType: file.type }));
```
Comment (stopfstedt) [inline packages/frontend/app/components/bulk-new-users.js:189]: this is an improvement over the current implementation (which throws an error as well, but it's not visible to the end user - only to be found in the browser console).

however it's not ideal:
image

File: packages/frontend/app/components/bulk-new-users.js [hunk 2]
```diff
@@ -234,18 +236,12 @@ export default class BulkNewUsersComponent extends Component {
...
-    const existingUsernames = await this.existingUsernames();
-    const filledOutUsers = proposedUsers.map((obj) => {
-      obj.existingUsernames = existingUsernames;
-
-      return obj;
-    });
-    this.validUsers = await filter(filledOutUsers, async (obj) => {
-      return await obj.isValid();
+    this.validUsers = await filter(proposedUsers, async (obj) => {
+      return obj.isValid();
```
Comment (stopfstedt) [inline packages/frontend/app/components/bulk-new-users.js:240]: `ProposedUser.isValid()` is async, so we should await for promise resolution here.

```suggestion
      return await obj.isValid();
```

File: packages/frontend/app/components/bulk-new-users.js (hunk 1 above)
Comment (michaelchadwick) [reply packages/frontend/app/components/bulk-new-users.js:189]: Interesting filename, that `DOOM PILG..er`. I'm guessing https://boardgamegeek.com/boardgame/373657/doom-pilgrim has been visited recently.

**Review by stopfstedt — CHANGES_REQUESTED** [review]
validation fails on a false positive here:

image

for  comparison from demo:
image

my test upload

```tsv
First	Last	Middle	Phone	Email	CampusID	OtherID	Username	Password
anton	zylinder	b	555-555-5555	anton@zylinder.com	xx9999991	xx9999991	xx9999991
anton	zylinder	b	notaphone	notanemail	xx9999992	xx9999992	xx9999992	
```

**Comment by jrjohnson** [comment]
> ```generic-db
> First	Last	Middle	Phone	Email	CampusID	OtherID	Username	Password
> anton	zylinder	b	555-555-5555	anton@zylinder.com	xx9999991	xx9999991	xx9999991
> ```

I *think* this bad data and was incorrectly not caught in the previous validation when I refactored it for octane 4 years ago. The password for this user is blank which isn't allowed if the username is present. However currently we do `const stringValue = String(this.password).trim();` in [the validation callback](https://github.com/ilios/frontend/blob/master/packages/frontend/app/components/bulk-new-users.js#L453) which doesn't work. When `this.password` is `null` `stringValue becomes "null"!! 🫣

I suppose the good news is that `"null"` is never gonna cause an issue if we store it as a password because we do a hash of the user input and no value will ever end up hashed to `"null"`, still though 🤦 

I will add a test for sure.

File: packages/frontend/app/components/bulk-new-users.js (hunk 2 above)
Comment (jrjohnson) [reply packages/frontend/app/components/bulk-new-users.js]: I'm always on the fence about this. Technically we're returning a promise into the `filter` and that filter will await the promise for us and act on the resolved value. So adding the await is unnecessary, but it does make it clearer what is happening.

File: packages/frontend/app/components/bulk-new-users.js (hunk 1 above)
Comment (michaelchadwick) [reply packages/frontend/app/components/bulk-new-users.js:189]: https://github.com/ilios/frontend/pull/8254 should help with avoiding such an error from even being generated.

**Review by stopfstedt — APPROVED** [review]
re-reviewed and click-tested. i couldn't find any regressions in the re-review. LGTM.

File: packages/frontend/app/components/bulk-new-users.js [hunk 3]
```diff
@@ -169,7 +168,7 @@ export default class BulkNewUsersComponent extends Component {
     this.primaryCohortId = id;
   }
 
-  async existingUsernames() {
+  async getExistingUserNames() {
```
Comment (michaelchadwick) [inline packages/frontend/app/components/bulk-new-users.js:171]: Do we not want to be consistent with naming convention here? I see `UserName`, but a few lines down on 184 you're using `Username`.

Comment (jrjohnson) [reply packages/frontend/app/components/bulk-new-users.js:171]: Damn it, I hate it when very good and valid points are made! Good catch. I will fix it.... tomorrow.

Comment (jrjohnson) [reply packages/frontend/app/components/bulk-new-users.js:171]: Fixed in #8258

**Pull request merged**